File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        istringstream iss(s);
        unordered_map<string,int> s2i;
        unordered_map<char,int> ch2i;
        int n {0} , sz = pattern.size();

        for(string word; iss >> word ; n++){
            if(n == sz || s2i[word] != ch2i[pattern[n]])
              return false;
            s2i[word] = ch2i[pattern[n]] = n + 1 ;
        }
        return n == sz;



    }
};

File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        s2i = {}
        ch2i = {}
        n = 0 
        sz = len(pattern)

        for word in words:
            if n == sz or s2i.get(word,-1) != ch2i.get(pattern[n],-1):
                return False
            s2i[word] = ch2i[pattern[n]] = n + 1
            n += 1
        return n == sz
        